### PR TITLE
[WasmFS] Support multiple readers in the OPFS backend

### PR DESCRIPTION
**Pull request by tlively**
Support multiple readers by using the Blob API instead of AccessHandles to
access files that are opened for reading only. This may come with a slight
performance penalty for accessing read-only files, but the previous behavior can
be recovered by opening file for both reading and writing.

When a file opened only for reading is additionally opened for writing, the
backend gracefully upgrades the access mode to use AccessHandles. Going the
other direction and downgrading the access mode back to Blobs when the last
writable file descriptor is closed is left as a future improvement.

Although it would be possible to support truncation of files opened in the Blob
access mode, we currently disallow it. Truncating (or otherwise modifying) a
file invalidates all Blobs that refer to it, so to support truncation we would
need to have a system for automatically refreshing Blobs for modified files.
That system would ideally also prevent concurrent in-flight operations on the
same file from observing the invalidated Blob.

For the same reason, if one tab writes to an OPFS file via AccessHandles, it
will invalidate the Blobs on all other tabs that have the file open for reading.
This will currently cause all subsequent reads on those other tabs to fail and
the only way to recover is for those tabs to close and reopen the affected
files. In the future it may be worth trying to automatically refresh the Blob
and retry when these errors occur, but in general it is not possible to prevent
the writer tab from starving out the reader tabs.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #17301** <a href="https://app.graphite.dev/github/pr/emscripten-core/emscripten/17301" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/emscripten-core/emscripten/17301?utm_source=stack-comment).

File: tests/wasmfs/wasmfs_opfs.c [hunk 1]
```diff
@@ -84,13 +85,6 @@ int main(int argc, char* argv[]) {
   assert(stat_buf.st_size == 100);
   emscripten_console_log("truncated to 100");
 
-  err = ftruncate(fd, 1);
-  assert(err == 0);
-  err = fstat(fd, &stat_buf);
-  assert(err == 0);
-  assert(stat_buf.st_size == 1);
-  emscripten_console_log("truncated to 1");
-
```
Comment (kripken) [inline tests/wasmfs/wasmfs_opfs.c:93]: Why was this part removed?

File: system/lib/wasmfs/backends/opfs_backend.cpp [hunk 2]
```diff
@@ -101,92 +114,221 @@ namespace {
...
+        return -EACCES;
+      }
+      // We have an AccessHandle, so close the blob.
+      proxy([&]() { _wasmfs_opfs_close_blob(getBlobID()); });
+      id = newID;
+      kind = Access;
+    } else {
+      // Reuse existing object.
+      ++openCount;
+      return 0;
```
Comment (kripken) [inline system/lib/wasmfs/backends/opfs_backend.cpp:173]: Does `openFlags` not need to be updated in this case? E.g. if this was opened for `WRONLY` and was now opened for `RDONLY` should the state not be `RDWR`?

File: system/lib/wasmfs/backends/opfs_backend.cpp [hunk 3]
```diff
@@ -101,92 +114,221 @@ namespace {
...
+        proxy([&](auto ctx) {
+          _wasmfs_opfs_set_size_access(ctx.ctx, state.getAccessID(), size);
+        });
+        break;
+      case OpenState::Blob:
+        // We don't support `truncate` in blob mode because the blob would
+        // become invalidated and refreshing it while ensuring other in-flight
+        // operations on the same file do not observe the invalidated blob would
+        // be extermely complicated.
+        assert(false && "TODO: proper error handling");
```
Comment (kripken) [inline system/lib/wasmfs/backends/opfs_backend.cpp:264]: I worry about depending on an assertion here, as if at some point we ship a build without assertions it would silently fall through, and the results could be very weird. How about `WASMFS_UNREACHABLE`?

File: system/lib/wasmfs/backends/opfs_backend.cpp [hunk 4]
```diff
@@ -101,92 +114,221 @@ namespace {
...
+      case OpenState::None: {
+        int err;
+        proxy([&](auto ctx) {
+          _wasmfs_opfs_set_size_file(ctx.ctx, fileID, size, &err);
+        });
+        assert(!err);
```
Comment (kripken) [inline system/lib/wasmfs/backends/opfs_backend.cpp:270]: Why is this guaranteed to succeed?

File: system/lib/wasmfs/syscalls.cpp [hunk 5]
```diff
@@ -492,7 +492,14 @@ static __wasi_fd_t doOpen(path::ParsedParent parsed,
   if (flags & O_TRUNC) {
     if (child->is<DataFile>()) {
       if (fileMode & WASMFS_PERM_WRITE) {
-        child->cast<DataFile>()->locked().setSize(0);
+        // TODO: Linux still truncates the file in when it's opened only for
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:495]: What does "in" refer to here?

File: system/lib/wasmfs/syscalls.cpp [hunk 6]
```diff
@@ -492,7 +492,14 @@ static __wasi_fd_t doOpen(path::ParsedParent parsed,
...
+        // reading, so we should too. But until we get proper error reporting
+        // from setSize (which we can then ignore here), the OPFS backend will
+        // crash in this case because it does not support truncating files open
+        // only for reading.
+        if (accessMode != O_RDONLY) {
+          child->cast<DataFile>()->locked().setSize(0);
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:501]: Could we put this workaround in the OPFS code and not in the generic syscall code? That is, OPFS setSize could handle it.

File: tests/wasmfs/wasmfs_opfs.c [hunk 7]
```diff
@@ -132,15 +126,59 @@ int main(int argc, char* argv[]) {
...
   close(fd);
 
   err = stat("/opfs/working/foo.txt", &stat_buf);
   assert(err == 0);
   assert(stat_buf.st_size == 0);
-  emscripten_console_log("statted after truncation");
+  emscripten_console_log("statted after successful truncation");
 
   err = rename("/opfs/working/foo.txt", "/opfs/foo.txt");
   assert(err == 0);
```
Comment (kripken) [inline tests/wasmfs/wasmfs_opfs.c:184]: It might be excessive, but perhaps it's worth a test that goes through all 9 combinations of opening in RDWR/WRONLY/RDONLY and then opens a second time in another mode?

File: system/lib/wasmfs/backends/opfs_backend.cpp (hunk 2 above)
Comment (tlively) [reply system/lib/wasmfs/backends/opfs_backend.cpp:173]: Ah, actually I noticed I can just get rid of `openFlags` entirely and use only `kind`...

File: system/lib/wasmfs/backends/opfs_backend.cpp (hunk 4 above)
Comment (tlively) [reply system/lib/wasmfs/backends/opfs_backend.cpp:270]: Heh, I don't think it is. I'll make this a `WASMFS_UNREACHABLE` as well.

File: system/lib/wasmfs/syscalls.cpp (hunk 5 above)
Comment (tlively) [reply system/lib/wasmfs/syscalls.cpp:495]: Oops, that's an extra word.

File: system/lib/wasmfs/syscalls.cpp (hunk 6 above)
Comment (tlively) [reply system/lib/wasmfs/syscalls.cpp:501]: That would require `setSize` to fail silently for now, but I would rather have this temporary hack and have `setSize` fail loudly for now than remove this hack and have `setSize` fail silently.

Comment (kripken) [reply system/lib/wasmfs/syscalls.cpp:501]: Oh, because `setSize` has no return value of an error code? That sounds good for now then with a TODO

Comment (tlively) [reply system/lib/wasmfs/syscalls.cpp:501]: Yeah, exactly. Is there any extra information you would want to see beyond the current TODO?

File: tests/wasmfs/wasmfs_opfs.c (hunk 7 above)
Comment (tlively) [reply tests/wasmfs/wasmfs_opfs.c:184]: Hmm yeah, that is probably warranted...

**Review by kripken — APPROVED** [review]
lgtm % comments

File: system/lib/wasmfs/syscalls.cpp (hunk 6 above)
Comment (tlively) [reply system/lib/wasmfs/syscalls.cpp:501]: I'm still not clear about what you'd like to see. There is already a TODO, but would you like to see an additional or modified TODO?

File: tests/wasmfs/wasmfs_opfs.c (hunk 1 above)
Comment (tlively) [reply tests/wasmfs/wasmfs_opfs.c:93]: It made the test nicer by not cutting off the message and we were already testing shortening files later. But I realize that the underlying state was different in the later shortening. This will be fixed by the more exhaustive testing.

File: system/lib/wasmfs/syscalls.cpp (hunk 6 above)
Comment (kripken) [reply system/lib/wasmfs/syscalls.cpp:501]: Sorry - I meant lgtm as it is,

File: tests/wasmfs/wasmfs_opfs.c (hunk 1 above)
Comment (kripken) [reply tests/wasmfs/wasmfs_opfs.c:93]: sgtm to leave for that testing later.
